euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/Sort.java

```diff
@@ -0,0 +1,345 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package cz.seznam.euphoria.core.client.operator;
+
+import cz.seznam.euphoria.core.annotation.operator.Derived;
+import cz.seznam.euphoria.core.annotation.operator.StateComplexity;
+import cz.seznam.euphoria.core.client.dataset.Dataset;
+import cz.seznam.euphoria.core.client.dataset.partitioning.HashPartitioning;
+import cz.seznam.euphoria.core.client.dataset.partitioning.Partitioner;
+import cz.seznam.euphoria.core.client.dataset.partitioning.Partitioning;
+import cz.seznam.euphoria.core.client.dataset.partitioning.RangePartitioning;
+import cz.seznam.euphoria.core.client.dataset.windowing.Window;
+import cz.seznam.euphoria.core.client.dataset.windowing.Windowing;
+import cz.seznam.euphoria.core.client.flow.Flow;
+import cz.seznam.euphoria.core.client.functional.UnaryFunction;
+import cz.seznam.euphoria.core.client.graph.DAG;
+import cz.seznam.euphoria.core.client.io.Context;
+import cz.seznam.euphoria.core.client.operator.state.ListStorage;
+import cz.seznam.euphoria.core.client.operator.state.ListStorageDescriptor;
+import cz.seznam.euphoria.core.client.operator.state.State;
+import cz.seznam.euphoria.core.client.operator.state.StateFactory;
+import cz.seznam.euphoria.core.client.operator.state.StorageProvider;
+import cz.seznam.euphoria.core.client.util.Pair;
+import cz.seznam.euphoria.shaded.guava.com.google.common.base.Preconditions;
+import cz.seznam.euphoria.shaded.guava.com.google.common.collect.Lists;
+import javax.annotation.Nullable;
+
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Sorts the input dataset.<p>
+ * 
+ * The user is supposed to provide a function that extracts a comparable object
+ * from the input object. The extracted object is than passed to a provided partitioner
+ * to partition the result.<p>
+ * 
+ * To ensure that all elements from a specific range end up 
+ * in the same partition - i.e. perform total sort ordering - it is recommended 
+ * to use {@link RangePartitioning}, otherwise the user is responsible for his partitioning.<p>
+ * 
+ * If user does not provide custom {@link Partitioning} or {@link Partitioner} and number of input 
+ * partitions differs from 1 (single partition), the program crashes, because runtime does not 
+ * know how to partition the result. Input sampling is not supported for now.<p> 
```

Comment:
Does it really crash? What is the reason not to use default `HashPartitioner` with same level of parallelism as input dataset? The result obviously wouldn't be totally ordered but it would be equal to Flink's "sort partition" transformation.